Grouping subpackage docstring updates

antalszava opened the pull request:
This PR contains minor adjustments to standardize the docstrings of the `grouping` subpackage. It mostly addresses formatting and standardization, without substantial content being added (hence separated from #850).

The changes include:

* Placed the `Example` section at the end of each docstring
* Renamed section headers `**Usage example:**`and `**Example usage:**` to `**Example**`
* Shortened variable names in some examples
* Updated the indentation for `Returns` section, removed names of returned objects, updated `tuple` return type docs
* Removed line break at the end of the docstrings
* Adjusted `Raises` in one case

The following is the way a `tuple` return type was documented:
```rst
tuple:
    
    * some_type1: some description or returned obj1
    * some_type2: some description or returned obj2
```
This renders as follows:
<a href=""><img src="https://user-images.githubusercontent.com/24476053/96199320-34dc6e00-0f25-11eb-96f7-8f388179ec5d.png" align="left" height="180" width="650" ></a>

antalszava commented on doc/code/qml_grouping.rst:
```diff
@@ -20,8 +20,8 @@ groupings according to a binary relation (e.g. qubit-wise commuting):
 
-    >>> observables = [qml.PauliY(0), qml.PauliX(0) @ qml.PauliX(1), qml.PauliZ(1)]
-    >>> obs_groupings = group_observables(observables)
+    >>> obs = [qml.PauliY(0), qml.PauliX(0) @ qml.PauliX(1), qml.PauliZ(1)]
```
Renaming vars

antalszava commented on pennylane/grouping/group_observables.py:
```diff
@@ -217,15 +197,29 @@ def group_observables(observables, coefficients=None, grouping_type="qwc", metho
+
+    **Example**
+
+    >>> obs = [qml.PauliY(0), qml.PauliX(0) @ qml.PauliX(1), qml.PauliZ(1)]
```
Renamed vars

antalszava commented on pennylane/grouping/optimize_measurements.py:
```diff
@@ -60,14 +42,30 @@ def optimize_measurements(observables, coefficients=None, grouping="qwc", colour
+
+    **Example**
+
+    >>> obs = [qml.PauliY(0), qml.PauliX(0) @ qml.PauliX(1), qml.PauliZ(1)]
```
Renamed vars

antalszava commented on pennylane/grouping/graph_colouring.py:
```diff
@@ -36,7 +36,17 @@ def largest_first(binary_observables, adj):
     number of vertices.
 
-    **Usage example:**
+    Args:
```
In such cases, the `Example` section was placed at the end of the docstring (GitHub diffs it odd)

josh146 commented on pennylane/grouping/group_observables.py:
```diff
@@ -188,22 +184,6 @@ def group_observables(observables, coefficients=None, grouping_type="qwc", metho
-    >>> coeffs_groupings
-    [[0.97, 4.21], [1.43]]
-
     Args:
         observables (list[Observable]): a list of Pauli word `Observable` instances (Pauli
             operation instances and Tensor instances thereof)
```
I can't comment directly as it is hidden, but we should move the named arguments below into `Args:`. We follow the convention in PL that `Keyword Args:` is reserved for arbitrary keyword args (e.g., `**kwargs`).

antalszava replied:
Thanks!

josh146 commented on pennylane/grouping/group_observables.py:
```diff
@@ -217,15 +197,29 @@ def group_observables(observables, coefficients=None, grouping_type="qwc", metho
+       tuple:
+
+           * list[list[Observable]]: A list of the obtained groupings. Each grouping
+             is itself a list of Pauli word `Observable` instances.
```
Also, there are a lot of other cases (that I can't comment on) where a single backtick was used instead of two.
```suggestion
             is itself a list of Pauli word ``Observable`` instances.
```

antalszava replied:
Thanks for catching them!

josh146 commented on pennylane/grouping/group_observables.py:
```diff
@@ -217,15 +197,29 @@ def group_observables(observables, coefficients=None, grouping_type="qwc", metho
             can be 'lf' (Largest First) or 'rlf' (Recursive Largest First)
```
I can't comment on all cases, but string options for parameters should be written

```suggestion
            can be ``'lf'`` (Largest First) or ``'rlf'`` (Recursive Largest First)
```

josh146 commented on pennylane/grouping/optimize_measurements.py:
```diff
@@ -62,14 +44,30 @@ def optimize_measurements(observables, coefficients=None, grouping="qwc", colour
-
+        tuple:
+
+            * list[Template]: a list of the post-rotation qml.Templates instances, one
+              for each partition
```
We don't yet have a `qml.Template` class 😆 
```suggestion
            * list[callable]: a list of the post-rotation templates, one
              for each partition
```

antalszava replied:
True, thank you! :smile:

josh146 commented on pennylane/grouping/optimize_measurements.py:
```diff
@@ -62,14 +44,30 @@ def optimize_measurements(observables, coefficients=None, grouping="qwc", colour
+            * list[list[Observable]]: A list of the obtained groupings. Each
+              grouping is itself a list of Pauli words diagonal in the
+              measurement basis.
+            * list[list[scalar]]: A list of coefficient groupings. Each
```
Hmm, we don't use 'scalar' in our docstrings. Should this be 
```suggestion
            * list[list[float]]: A list of coefficient groupings. Each
```

antalszava replied:
Fair point, thanks!

josh146 commented on pennylane/grouping/group_observables.py:
```diff
@@ -217,15 +197,29 @@ def group_observables(observables, coefficients=None, grouping_type="qwc", metho
+
+           * list[list[Observable]]: A list of the obtained groupings. Each grouping
+             is itself a list of Pauli word `Observable` instances.
+           * list[list[scalar]]: A list of coefficient groupings. Each coefficient
```
```suggestion
           * list[list[float]]: A list of coefficient groupings. Each coefficient
```

josh146 commented on pennylane/grouping/graph_colouring.py:
```diff
@@ -36,7 +36,17 @@ def largest_first(binary_observables, adj):
 
-    **Usage example:**
+    Args:
+        binary_observables (array[bool]): the set of Pauli words represented by a column matrix
+            of the Pauli words in binary vector represenation
+        adj (array[bool]): the adjacency matrix of the Pauli graph
+
+    Returns:
+        dict(int, list[array[bool]]): keys correspond to colours (labelled by integers) and values
```
From the examples below, this should be `float`. Although, it would probably make more sense for these to be int? Unless weighted adjacency matrices are allowed...
```suggestion
        binary_observables (array[float]): the set of Pauli words represented by a column matrix
            of the Pauli words in binary vector represenation
        adj (array[float]): the adjacency matrix of the Pauli graph

    Returns:
        dict(int, list[array[float]]): keys correspond to colours (labelled by integers) and values
```

antalszava replied:
Thanks for catching that Josh! Think it could indeed only be int (binary for Paulis & adjacency without weights)

josh146 commented on pennylane/grouping/graph_colouring.py:
```diff
@@ -91,7 +89,16 @@ def recursive_largest_first(binary_observables, adj):  # pylint:disable=too-many
 
-    **Usage example:**
+    Args:
+        binary_observables (array[bool]): the set of Pauli words represented by a column matrix of
+            the Pauli words in binary vector represenation
+        adj (array[bool]): the adjacency matrix of the Pauli graph
```
Same here re: bool

josh146 reviewed (commented):
Thanks for this PR @antalszava! Unfortunately, I noticed a few rendering issues and odd conventions still present.

* There are quite a few cases of single backticks instead of double backticks

* Named arguments with defaults should go in `Args:`, not `Keyword Args:`

* Odd or incorrect types (e.g., `array[bool]`, `scalar`, `qml.Template`).

Unfortunately, I was not able to comment on all of them, since some were not showing in the FIles Changed tab!

antalszava commented:
Thank you for the comments @josh146!

Should have addressed them all (mostly used search & replace to catch the occurrences).

In specific:
* Corrected single backticks to double backticks
* Moved named arguments with default into `Args` (from `Keyword Args`)
* Made type adjustments:
   - `array[bool]` -> `array[int]`
   - `scalar` -> `float`
   - `qml.Template` -> `callable`

antalszava commented on pennylane/grouping/group_observables.py:
```diff
@@ -49,16 +49,13 @@ class PauliGroupingStrategy:  # pylint: disable=too-many-instance-attributes
     Args:
         observables (list[Observable]): a list of Pauli words to be partitioned according to a
             grouping strategy
-
-    Keyword Args:
```
`Keyword Args:` section was removed such that named arguments with defaults go in `Args:` (recurring change)

co9olguy commented on pennylane/grouping/graph_colouring.py:
```diff
@@ -91,7 +88,16 @@ def recursive_largest_first(binary_observables, adj):  # pylint:disable=too-many
+
+    Returns:
+        dict(int, list[array[int]]): keys correspond to colours (labelled by integers) and values
+        are lists of Pauli words of the same colour in binary vector representation.
```
```suggestion
        are lists of Pauli words of the same colour in binary vector representation
```

co9olguy commented on pennylane/grouping/utils.py:
```diff
@@ -117,11 +116,26 @@ def pauli_to_binary(pauli_word, n_qubits=None, wire_map=None):
+                         unique integer labels as their values
+
+    Returns:
+        array: the 2*n_qubits dimensional binary vector representation of the input Pauli word.
```
```suggestion
        array: the 2*n_qubits dimensional binary vector representation of the input Pauli word
```

co9olguy reviewed (approved):
Thanks @antalszava, the changes you've made look good.

Note that I did not carefully check if there were further changes that could be done

josh146 reviewed (approved):
💯
